File: listas/src/main/java/com/victorlh/android/framework/listas/adapter/AbstractLista.java
package com.victorlh.android.framework.listas.adapter;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.CallSuper;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.RecyclerView;

import com.victorlh.android.framework.listas.ItemLista;
import com.victorlh.android.framework.listas.R;
import com.victorlh.tools.ordenacion.ElementoOrdenado;
import com.victorlh.tools.ordenacion.KeyOrdenacion;
import com.victorlh.tools.ordenacion.UtilOrdenacion;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;

/**
 * Created by victor on 07/07/2017.
 */
public abstract class AbstractLista<T extends ItemLista> extends RecyclerView.Adapter<AbstractViewHolder<T>> implements Lista<T> {

	private Class<T> clazz;

	private List<T> lista;

	private HashMap<T, AbstractViewHolder<T>> viewHolders;

	private OnClickElementoListener<T> onClickElementoListener;
	private OnLongClickElementoListener<T> onLongClickElementoListener;
	private OnDataListChangeListener onDataListChangeListener;

	private KeyOrdenacion keyOrdenacion;

	private ListaObservableAdapter<T> observableAdapter;

	protected AbstractLista(Class<T> clazz) {
		this.clazz = clazz;

		this.lista = new ArrayList<>();
		this.viewHolders = new HashMap<>();

		this.observableAdapter = new ListaObservableAdapter<>();
	}

	public void setKeyOrdenacion(KeyOrdenacion keyOrdenacion) {
		this.keyOrdenacion = keyOrdenacion;
	}

	protected abstract AbstractViewHolder<T> createViewHolder(View view);

	@NonNull
	@Override
	public AbstractViewHolder<T> onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
		Context context = parent.getContext();

		LayoutInflater inflater = LayoutInflater.from(context);
		View itemView = inflater.inflate(R.layout.item_listas_wrapper, parent, false);

		AbstractViewHolder<T> viewHolder = createViewHolder(itemView);
		viewHolder.setLista(this);
		itemView.setTag(viewHolder);
		return viewHolder;
	}

	@Override
	@CallSuper
	public void onBindViewHolder(@NonNull final AbstractViewHolder<T> holder, int position) {
		OnClickListaAdapter listener = new OnClickListaAdapter(holder);
		holder.setOnClickListaAdapter(listener);

		final T item = getItem(position);
		viewHolders.put(item, holder);
		holder.procesar(item);
		observableAdapter.onBindViewHolder(holder, position);
	}

	@Override
	public long getItemId(int position) {
		T item = getItem(position);
		return item == null ? super.getItemId(position) : item.getId();
	}

	@Override
	public List<T> getLista() {
		return new ArrayList<>(lista);
	}

	@Override
	public T getItem(int position) {
		if (position < 0 || position > getItemCount()) {
			return null;
		}
		return lista.get(position);
	}

	@Override
	public int getItemCount() {
		return lista.size();
	}

	@Override
	public int getItemPosition(T item) {
		return getItemPosition(item, false);
	}

	@Override
	public int getItemPosition(T item, boolean byId) {
		int i = lista.indexOf(item);
		if (i == -1 && byId) {
			long id = item.getId();
			T aux = getItemById(id);
			if (aux != null) {
				i = lista.indexOf(aux);
			}
		}
		return i;
	}

	@Override
	public T getItemById(long id) {
		for (T item : lista) {
			if (item.getId() == id) {
				return item;
			}
		}
		return null;
	}

	@Override
	public void setLista(T[] lista) {
		List<T> ts = Arrays.asList(lista);
		setLista(ts);
	}

	@Override
	@SuppressWarnings("unchecked")
	public void setLista(List<T> lista) {
		if (keyOrdenacion != null && ElementoOrdenado.class.isAssignableFrom(clazz)) {
			UtilOrdenacion.ordenar((List<? extends ElementoOrdenado>) lista, keyOrdenacion);
		}
		boolean empty = this.lista.isEmpty();
		this.lista.clear();
		this.lista.addAll(lista);
		recargarLista();
		if (onDataListChangeListener != null) {
			if (empty && !this.lista.isEmpty()) {
				onDataListChangeListener.onSetData();
			} else if (!empty && this.lista.isEmpty()) {
				onDataListChangeListener.onVoidData();
			} else {
				onDataListChangeListener.onChangeData();
			}
		}
	}

	@Override
	@SuppressWarnings("unchecked")
	public void addItem(T item) {
		if (keyOrdenacion != null && ElementoOrdenado.class.isAssignableFrom(clazz)) {
			List<T> aux = getLista();
			aux.add(item);
			UtilOrdenacion.ordenar((List<? extends ElementoOrdenado>) aux, keyOrdenacion);
			int position = aux.indexOf(item);
			addItem(item, position);
		} else {
			addItem(item, this.lista.size());
		}
	}

	@Override
	public void addItem(T item, int position) {
		boolean empty = this.lista.isEmpty();
		if (position == lista.size()) {
			lista.add(item);
		} else {
			lista.add(position, item);
		}
		notifyItemInserted(position);
		if (onDataListChangeListener != null) {
			if (empty) {
				onDataListChangeListener.onSetData();
			} else {
				onDataListChangeListener.onChangeData();
			}
		}
	}

	@Override
	@SuppressWarnings("unchecked")
	public void setItem(T item) {
		int itemPosition = getItemPosition(item, true);
		if (itemPosition == -1) {
			addItem(item);
		} else {
			lista.set(itemPosition, item);
			notifyItemChanged(itemPosition);
			if (keyOrdenacion != null && ElementoOrdenado.class.isAssignableFrom(clazz)) {
				List<T> auxList = getLista();
				UtilOrdenacion.ordenar((List<? extends ElementoOrdenado>) auxList, keyOrdenacion);
				int destino = auxList.indexOf(item);
				if (itemPosition != destino) {
					this.lista = auxList;
					notifyItemMoved(itemPosition, destino);
				}
			}
			observableAdapter.setItem(item);
			if (onDataListChangeListener != null) {
				onDataListChangeListener.onChangeData();
			}
		}
	}

	@Override
	public void deleteItem(T item) {
		int itemPosition = getItemPosition(item, true);
		if (itemPosition != -1) {
			lista.remove(itemPosition);
			notifyItemRemoved(itemPosition);
			observableAdapter.deleteItem(item);
			if (onDataListChangeListener != null && lista.isEmpty()) {
				onDataListChangeListener.onVoidData();
			}
		}
	}

	@Override
	public void recargarLista() {
		this.viewHolders.clear();
		notifyDataSetChanged();
		observableAdapter.recargarLista();
	}

	@Override
	public void setOnClickElementoListener(OnClickElementoListener<T> onClickElementoListener) {
		this.onClickElementoListener = onClickElementoListener;
	}

	@Override
	public void setOnLongClickElementoListener(OnLongClickElementoListener<T> onLongClickElementoListener) {
		this.onLongClickElementoListener = onLongClickElementoListener;
	}

	@Override
	public void setOnDataListChangeListener(OnDataListChangeListener onDataListChangeListener) {
		this.onDataListChangeListener = onDataListChangeListener;
	}

	@Nullable
	public final AbstractViewHolder<T> getViewHolder(T item) {
		return viewHolders.get(item);
	}

	public Class<T> getClazz() {
		return clazz;
	}

	public void registrarObservable(ListaObservable<T> observable) {
		observableAdapter.registrarObservable(observable);
	}

	public void removerObservable(ListaObservable<T> observable) {
		observableAdapter.removerObservable(observable);
	}

	class OnClickListaAdapter implements View.OnClickListener, View.OnLongClickListener {

		protected final AbstractViewHolder<T> holder;

		private OnClickListaAdapter(AbstractViewHolder<T> holder) {
			this.holder = holder;
		}

		@Override
		public void onClick(View v) {
			observableAdapter.onClickElemento(holder);
			if (onClickElementoListener != null) {
				onClickElementoListener.onClickElemento(holder);
			}
		}

		@Override
		public boolean onLongClick(View v) {
			observableAdapter.onLongClickElemento(holder);
			if (onLongClickElementoListener != null) {
				return onLongClickElementoListener.onLongClickElemento(holder);
			}
			return false;
		}
	}
}

File: listas/src/main/java/com/victorlh/android/framework/listas/adapter/AbstractLista.kt
package com.victorlh.android.framework.listas.adapter

import android.view.LayoutInflater
import android.view.View
import android.view.View.OnLongClickListener
import android.view.ViewGroup
import androidx.annotation.CallSuper
import androidx.recyclerview.widget.RecyclerView
import com.victorlh.android.framework.listas.ItemLista
import com.victorlh.android.framework.listas.R
import com.victorlh.android.framework.listas.adapter.Lista.*
import com.victorlh.tools.ordenacion.ElementoOrdenado
import com.victorlh.tools.ordenacion.KeyOrdenacion
import com.victorlh.tools.ordenacion.UtilOrdenacion

/**
 * Created by victor on 07/07/2017.
 */
abstract class AbstractLista<T> protected constructor(val clazz: Class<T>) : RecyclerView.Adapter<AbstractViewHolder<T>>(), Lista<T> {

    private var lista: MutableList<T> = ArrayList()
    private val viewHolders: HashMap<T, AbstractViewHolder<T>> = HashMap()
    private val observableAdapter: ListaObservableAdapter<T> = ListaObservableAdapter()

    private var onClickElementoListener: OnClickElementoListener<T>? = null
    private var onLongClickElementoListener: OnLongClickElementoListener<T>? = null
    private var onDataListChangeListener: OnDataListChangeListener? = null
    private var keyOrdenacion: KeyOrdenacion? = null

    fun setKeyOrdenacion(keyOrdenacion: KeyOrdenacion?) {
        this.keyOrdenacion = keyOrdenacion
    }

    protected abstract fun createViewHolder(view: View?): AbstractViewHolder<T>

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): AbstractViewHolder<T> {
        val context = parent.context
        val inflater = LayoutInflater.from(context)
        val itemView = inflater.inflate(R.layout.item_listas_wrapper, parent, false)
        val viewHolder = createViewHolder(itemView)
        viewHolder.lista = this
        itemView.tag = viewHolder
        return viewHolder
    }

    @CallSuper
    override fun onBindViewHolder(holder: AbstractViewHolder<T>, position: Int) {
        val listener = OnClickListaAdapter(holder)
        holder.setOnClickListaAdapter(listener)

        val item = getItem(position)
        viewHolders[item] = holder
        holder.procesar(item)
        observableAdapter.onBindViewHolder(holder, position)
    }

    override fun getItemId(position: Int): Long {
        val item = getItem(position) ?: return super.getItemId(position)
        return getItemId(item)
    }

    override fun getLista(): MutableList<T> {
        return ArrayList(lista)
    }

    override fun getItem(position: Int): T {
        return lista[position]
    }

    override fun getItemCount(): Int {
        return lista.size
    }

    override fun getItemPosition(item: T): Int {
        return getItemPosition(item, false)
    }

    override fun getItemPosition(item: T, byId: Boolean): Int {
        var i = lista.indexOf(item)
        if (i == -1 && byId) {
            val id = getItemId(item)
            val aux: T? = getItemById(id)
            if (aux != null) {
                i = lista.indexOf(aux)
            }
        }
        return i
    }

    override fun getItemById(id: Long): T? {
        for (item in lista) {
            if (getItemId(item) == id) {
                return item
            }
        }
        return null
    }

    override fun setLista(lista: Array<T>) {
        val ts = listOf(*lista)
        setLista(ts)
    }

    @Suppress("UNCHECKED_CAST")
    override fun setLista(lista: List<T>) {
        if (keyOrdenacion != null && ElementoOrdenado::class.java.isAssignableFrom(clazz)) {
            UtilOrdenacion.ordenar(lista as List<ElementoOrdenado>, keyOrdenacion)
        }
        val empty = this.lista.isEmpty()
        this.lista.clear()
        this.lista.addAll(lista)
        recargarLista()
        if (onDataListChangeListener != null) {
            if (empty && this.lista.isNotEmpty()) {
                onDataListChangeListener!!.onSetData()
            } else if (!empty && this.lista.isEmpty()) {
                onDataListChangeListener!!.onVoidData()
            } else {
                onDataListChangeListener!!.onChangeData()
            }
        }
    }

    @Suppress("UNCHECKED_CAST")
    override fun addItem(item: T) {
        if (keyOrdenacion != null && ElementoOrdenado::class.java.isAssignableFrom(clazz)) {
            val aux = getLista()
            aux.add(item)
            UtilOrdenacion.ordenar(aux as List<ElementoOrdenado>, keyOrdenacion)
            val position = aux.indexOf(item)
            addItem(item, position)
        } else {
            addItem(item, lista.size)
        }
    }

    override fun addItem(item: T, position: Int) {
        val empty = lista.isEmpty()
        if (position == lista.size) {
            lista.add(item)
        } else {
            lista.add(position, item)
        }
        notifyItemInserted(position)
        if (onDataListChangeListener != null) {
            if (empty) {
                onDataListChangeListener!!.onSetData()
            } else {
                onDataListChangeListener!!.onChangeData()
            }
        }
    }

    @Suppress("UNCHECKED_CAST")
    override fun setItem(item: T) {
        val itemPosition = getItemPosition(item, true)
        if (itemPosition == -1) {
            addItem(item)
        } else {
            lista[itemPosition] = item
            notifyItemChanged(itemPosition)
            if (keyOrdenacion != null && ElementoOrdenado::class.java.isAssignableFrom(clazz)) {
                val auxList = getLista()
                UtilOrdenacion.ordenar(auxList as List<ElementoOrdenado?>, keyOrdenacion)
                val destino = auxList.indexOf(item)
                if (itemPosition != destino) {
                    lista = auxList
                    notifyItemMoved(itemPosition, destino)
                }
            }
            observableAdapter.setItem(item)
            if (onDataListChangeListener != null) {
                onDataListChangeListener!!.onChangeData()
            }
        }
    }

    override fun deleteItem(item: T) {
        val itemPosition = getItemPosition(item, true)
        if (itemPosition != -1) {
            lista.removeAt(itemPosition)
            notifyItemRemoved(itemPosition)
            observableAdapter.deleteItem(item)
            if (onDataListChangeListener != null && lista.isEmpty()) {
                onDataListChangeListener!!.onVoidData()
            }
        }
    }

    override fun recargarLista() {
        viewHolders.clear()
        notifyDataSetChanged()
        observableAdapter.recargarLista()
    }

    override fun setOnClickElementoListener(onClickElementoListener: OnClickElementoListener<T>) {
        this.onClickElementoListener = onClickElementoListener
    }

    override fun setOnLongClickElementoListener(onLongClickElementoListener: OnLongClickElementoListener<T>) {
        this.onLongClickElementoListener = onLongClickElementoListener
    }

    override fun setOnDataListChangeListener(onDataListChangeListener: OnDataListChangeListener) {
        this.onDataListChangeListener = onDataListChangeListener
    }

    fun getViewHolder(item: T): AbstractViewHolder<T>? {
        return viewHolders[item]
    }

    fun registrarObservable(observable: ListaObservable<T>?) {
        observableAdapter.registrarObservable(observable!!)
    }

    fun removerObservable(observable: ListaObservable<T>?) {
        observableAdapter.removerObservable(observable)
    }

    internal fun getItemId(item: T): Long {
        return if (item is ItemLista) (item as ItemLista).id else item.hashCode().toLong()
    }

    internal inner class OnClickListaAdapter constructor(private val holder: AbstractViewHolder<T>) : View.OnClickListener, OnLongClickListener {
        override fun onClick(v: View) {
            val interceptado = observableAdapter.onClickElemento(holder)
            if (!interceptado && onClickElementoListener != null) {
                onClickElementoListener!!.onClickElemento(holder)
            }
        }

        override fun onLongClick(v: View): Boolean {
            val interceptObservable = observableAdapter.onLongClickElemento(holder)
            return if (!interceptObservable && onLongClickElementoListener != null) {
                onLongClickElementoListener!!.onLongClickElemento(holder)
            } else interceptObservable
        }

    }
}